### ISSUE TITLE
Proposal for prepared statements

### ISSUE THREAD

**Issue by jhannus** [cequel/cequel#306]
Hello Cequel Developers,

I'm hoping to start a discussion about the possibility of adding prepared statements to the cequel gem. I'm thinking that at a high-level prepared statements could be introduced (possibly in this order):

- The finder methods generated for each model by cequel. These seem like the best candidate and the lowest hanging fruit since they are already auto generated and are single use case driven.
- Associations `has_many` and `belongs_to` (for similar reasons as above)
- #save, #update, and #delete
- Custom queries i.e. `Model.select(:foo).where(:condition).prepared` where `prepared` tells cequel to lately bind a prepared statement?
- Usage in batches?

I'm thinking that maybe a configuration option would tell cequel to use prepared statements for the auto generated finders, associations, and save methods where the `prepared` method would always be available to invoke on relation objects.

Thoughts?

Thanks!
-j

**Comment by pezra** [comment]
Awesome idea. We should be using prepared statements where possible.

I'd be inclined to have prepared statements be the only way finders and associations work. Are there downsides to using prepared statements that necessitate an option to revert to the dynamic statement behavior?

It seems like it might be easier to start with custom queries because it is more isolated.  However, what to do first is a decision best left to whoever decides to put together the first PR. :)

**Comment by jhannus** [comment]
> Are there downsides to using prepared statements...
None that I can think of aside from some edge-case-like stuff. It just seems like some folks might want the option? Maybe this piece can be omitted from start. If needed it could be added later on.

> It seems like it might be easier to start with custom queries because it is more isolated
Good point. I could imagine a single mechanism is responsible for locating prepared statements before executing raw CQL so that finders, relations, and custom queries can technically be done in any order with that affordance.

I'm trying to get this planned for my next sprint. I'll let you know how it goes.

👍

**Comment by pezra** [comment]
Sounds good. I look forward to seeing the PR. :smile:

**Comment by jhannus** [comment]
tldr;
I have a basic working version of prepared statements for the above proposal including use within batches. The amount of refactoring is significant and the code is not ready. I'll explain below :)

Ok @pezra I have a _decent_ working version of this in my internal Github which I plan on cleaning up a little bit and moving to my public Github fork for early review. Before doing so I'd like to call out a few things because it's not finished enough for an open source contribution/merge but is enough for my apps use cases and testing.

First there was some significant refactoring needed to support both prepared and non-prepared queries. Or rather, to even support prepared... you **need** to use bound parameters. The Cequel gem not only does all CQL generation but it also does its own bind parameter quoting and replacement with all strings before sending off to the drivers `#execute()` method. This by definition cannot be prepared because the values would already be in the CQL. I learned this after getting into deep into the code.

In the case of batches the CQL statement (described above) is concatenated for each query in the batch with their respective params replaced resulting in one large BATCH statement as a string sent to the drivers `#execute()` as well.

So I went down the path of refactoring `Cequel::Metal::Statement` to be a single statement with only that single statements bind vars. I then refactored `Cequel::Metal::Batch` to create an array of these `Statements` so that when `Cequel::Metal::Batch#apply` is called it could create a `Cassandra::Statements::Batch` instance and add each statement separately with the arguments. See: http://datastax.github.io/ruby-driver/features/basics/batch_statements/ for an example. This allows for the CQL to be prepared or non-prepared with the same argument list.

Similar for executing a single statement (not in batch).

There are 2 issues that are catching me up:
* Collection add/remove operations need some more work. We don't use these in our app.
* When using bind parameters for non-prepared statements the ruby driver needs type hints on certain types of columns or it will produce a `Cassandra::Errors::InvalidError`.

In order to fully support prepared statements in the Cequel gem these 2 issues need to be resolved. I have some ideas for the type issues and came up with some examples but need to further refactor.

Very doable but some more work is required.

**Comment by pezra** [comment]
It is great to hear of the progress. Prepared statement support will be an excellent addition to Cequel.

Your troubles with bound parameters are familiar. PR #311 was an attempt to accomplish that but it did not reach completion. @dividedmind, are the issues above similar to the issues you encountered?

@jhannus, would it be reasonable to divide the work by putting together a bound parameter PR? Once that was merged it sounds like the prepared statement refactor would be easier. If you extracted the bound parameter refactors you have and put them in a PR i'd be happy to help out with the finishing touches.

**Comment by jhannus** [comment]
That sounds reasonable. I kept my history in stages and it will take some work but I could come up with a PR for bound parameters with batch support, then come up with a patch for the prepared support.

**Comment by jhannus** [comment]
I say "with batch" support meaning cassandra-driver native batch support which will set us up well for the prepared statements.

**Comment by pezra** [comment]
Should using the driver's native batch capabilities be a separate PR?

**Comment by jhannus** [comment]
Yeah thats possible. Maybe there are 3 PR's with the statement refactor for native batch being the first? Then bound followed by prepared. 

The prepared support (which is what we want) cannot happen with batches being a single statement. Once thats in place the bound params will need type hints which will be easier to manage per statement object as opposed to large arrays for one.

**Comment by jhannus** [comment]
I pulled out the statements refactor to prep for the bound params. You can find the commit here: https://github.com/jhannus/cequel/commit/4974c20df91ae8b5a2809d4c3287ae8a828d0dfd

The Statement class is now only used for single statements and encapsulates its own bind variables and type hints. The batch now uses the Cassandra native batch support. 

I only changed the `execute_with_consistency` to use `execute_with_options` because thats where the type hints need go and there needs to be a way to pass them from the Writers and the DataSet. I removed the method since I didn't see it used anywhere else.

The type hints are needed for bound params even for simple int columns but **only** when _not_ prepared. The reason for this is the driver cannot distinguish between ints or bigints from ruby's Fixnum type. The same goes for floats, doubles, and collections of. All scripting drivers have this issue (node, php, ruby). Somehow using prepared statements it can auto figure the type, however. 

I have an open line with the DataStax guys for the project I'm woking on now and we went over this exact issue and there is no plan to fix. See: https://datastax-oss.atlassian.net/browse/RUBY-15 for more details.

**Comment by jhannus** [comment]
Sorry for my late reply but I've been on vacation and school has just started for the kiddos 😄  I've also switched over to doing this in my spare time.

Here you can find the larger diff which includes everything except cleanup/rebase/etc. There are a few failing specs but they're mostly due to the type hints not being present: https://github.com/jhannus/cequel/compare/master...jhannus:prepared-batch-support?expand=1

And here are the two separated commits of the statement refactor @pezra and I talked about above. The type hinting was only applied to some of the writers. I'm not exactly sure I like the way the type hints are being passed around so maybe I could get some feedback/suggestions on it before continuing? https://github.com/jhannus/cequel/compare/master...jhannus:statements-refactor?expand=1

**Comment by jhannus** [comment]
**Edit:** I have an update to that second compare link: https://github.com/cequel/cequel/compare/master...jhannus:statements-refactor?expand=1

**Comment by pezra** [comment]
I think that branch is coming along nicely. Does the user normally have to provide explicit type hints when using that branch? Would it be better to create a PR out of it to get a few more eyes on the issue?

**Comment by jhannus** [comment]
No, the user should only have to define their tables in the model just as they do today. Before a statement is executed, during cql generation, it should read from the table's column definitions to get the types.

This is sort of in place already but I'm not quite happy with the way I first tried it. Since the writers are responsible for cql generation that seems like a good place it's just the accessors to the table object I found a lil troubling.

Sure I can open a PR. I'm in San Jose for the week for the Cassandra Summit. I'll try to find some time. In the meantime, anyone else going to be here at the conference?
